File: cpp/binary_tree_level_order_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // simplified dfs array approach
    vector<vector<int>> levelOrder(TreeNode *root) {
        vector<vector<int>> nodesByLevel;
        encodeNodes(root, 0, nodesByLevel);
        return nodesByLevel;
    }

    void encodeNodes(TreeNode *curNode, int curLevel, vector<vector<int>> &nodesByLevel) {
        if (curNode == nullptr) {
            return;
        }

        // observation: any recursive dfs order will always visit all
        // nodes across a level in order

        // extend to store node values at current level upon first time reaching level
        if (nodesByLevel.size() < curLevel + 1) {
            nodesByLevel.push_back({});
        }
        nodesByLevel[curLevel].push_back(curNode->val);

        // collect and write all nodes underneath current level
        encodeNodes(curNode->left, curLevel + 1, nodesByLevel);
        encodeNodes(curNode->right, curLevel + 1, nodesByLevel);
    }

    // // dfs hashmap approach
    // vector<vector<int>> levelOrder(TreeNode *root) {
    //     map<int, vector<int>> nodesByLevel;
    //     encodeNodes(root, 0, nodesByLevel);

    //     vector<vector<int>> answer;
    //     for (auto &[_, nodeLevel] : nodesByLevel) {
    //         answer.push_back(nodeLevel);
    //     }
    //     return answer;
    // }

    // void encodeNodes(TreeNode *curNode, int curLevel, map<int, vector<int>> &nodesByLevel) {
    //     if (curNode == nullptr) {
    //         return;
    //     }

    //     // collect and write all nodes underneath current level first
    //     encodeNodes(curNode->left, curLevel + 1, nodesByLevel);
    //     encodeNodes(curNode->right, curLevel + 1, nodesByLevel);
    //     nodesByLevel[curLevel].push_back(curNode->val);
    // }

    // // bfs approach
    // vector<vector<int>> levelOrder(TreeNode* root) {
    //     if (root == nullptr) {
    //         return {};
    //     }

    //     vector<vector<int>> nodesByLevel;
    //     deque<TreeNode*> nodesToVisit = { root };

    //     while (!nodesToVisit.empty()) {
    //         int totalLevelNodes = nodesToVisit.size();
    //         vector<int> nodeLevel;

    //         // capture and store all node values in current bfs level
    //         for (int _ = 0; _ < totalLevelNodes; ++_) {
    //             TreeNode *curNode = nodesToVisit.front();
    //             nodesToVisit.pop_front();
    //             nodeLevel.push_back(curNode->val);

    //             // do not queue null children to visit in next level
    //             if (curNode->left != nullptr) {
    //                 nodesToVisit.push_back(curNode->left);
    //             }
    //             if (curNode->right != nullptr) {
    //                 nodesToVisit.push_back(curNode->right);
    //             }
    //         }

    //         nodesByLevel.push_back(nodeLevel);
    //     }

    //     return nodesByLevel;
    // }
};

File: python/binary_tree_level_order_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
#     dfs preorder traversal using map
#     def levelOrder(self, root: TreeNode) -> List[List[int]]:
#         childrenAtDepthLookup: Dict[int, List[int]] = {}

#         self.organizeChildrenByDepth(root, childrenAtDepthLookup, 0)
#         childrenAtDepth = []

#         for depth in childrenAtDepthLookup:
#             childrenAtDepth.append(childrenAtDepthLookup[depth])

#         return childrenAtDepth

#     def organizeChildrenByDepth(self, root: TreeNode, childrenAtDepthLookup: Dict[int, List[int]], currDepth: int):
#         if root == None:
#             return
#
#         if currDepth not in childrenAtDepthLookup:
#             childrenAtDepthLookup[currDepth] = []

#         childrenAtDepthLookup[currDepth].append(root.val)

#         self.organizeChildrenByDepth(root.left, childrenAtDepthLookup, currDepth + 1)
#         self.organizeChildrenByDepth(root.right, childrenAtDepthLookup, currDepth + 1)

#     # cleaner dfs write to list by recursion depth index
#     def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
#         nodesByLevel: List[List[int]] = []
#         self.encodeNodesByLevel(root, nodesByLevel, 0)
#         return nodesByLevel

#     def encodeNodesByLevel(self, root: Optional[TreeNode], nodesByLevel: List[List[int]], levelIndex: int):
#         if not root:
#             return

#         if len(nodesByLevel) <= levelIndex:
#             nodesByLevel.append([])

#         nodesByLevel[levelIndex].append(root.val)

#         self.encodeNodesByLevel(root.left, nodesByLevel, levelIndex + 1)
#         self.encodeNodesByLevel(root.right, nodesByLevel, levelIndex + 1)

    # bfs approach, queue each level's nodes and copy to result
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        nodesToVisit = collections.deque([root])
        nodesByLevel = []

        while nodesToVisit:
            # take size snapshot of current level
            levelSize = len(nodesToVisit)
            # copy level's node values to result set
            nodeLevelValues = []

            # replace current level up to snapshot size with next level's nodes
            for _ in range(levelSize):
                curNode = nodesToVisit.popleft()
                nodeLevelValues.append(curNode.val)

                if curNode.left:
                    nodesToVisit.append(curNode.left)
                if curNode.right:
                    nodesToVisit.append(curNode.right)

            nodesByLevel.append(nodeLevelValues)

        return nodesByLevel
